invalid signature during block production on altair minimal network

scenario: started a minimal network of altair to do some testing for APIs, and noticed this a couple of times in my logs:
```
2021-05-07 13:30:51.025+10:00 | validator-async-8 | ERROR | teku-validator-log | ^[[31mValidator   *** Failed to produce block  Slot: 832 Validator: ae00fc3^[[0m
tech.pegasys.teku.spec.logic.common.statetransition.exceptions.StateTransitionException: tech.pegasys.teku.spec.logic.common.statetransition.exceptions.BlockProcessingException: Invalid sync committee signature in SyncAggregate{sync_committee_bits=SszBitvector{size=32, 11111111111111111111111111111111}, sync_committee_signature=SszByteVector{0x8a68f4032257182582d02afe12569a461296d06dc85115d504ac2929e6b277405f3bbac0e730274eadca97747d7ef899030f7f9033efb746dae154c51c1a9a4f088438db8ab15fcf03620d301fc5d21c0ca2af08e8a7d9fafe9864ec8640b318}}
        at tech.pegasys.teku.spec.logic.common.util.BlockProposalUtil.createNewUnsignedBlock(BlockProposalUtil.java:80) ~[classes/:?]
        at tech.pegasys.teku.spec.Spec.createNewUnsignedBlock(Spec.java:414) ~[classes/:?]
        at tech.pegasys.teku.validator.coordinator.BlockFactory.createUnsignedBlock(BlockFactory.java:122) ~[classes/:?]
        at tech.pegasys.teku.validator.coordinator.ValidatorApiHandler.createBlock(ValidatorApiHandler.java:309) ~[classes/:?]
        at tech.pegasys.teku.validator.coordinator.ValidatorApiHandler.lambda$createUnsignedBlock$12(ValidatorApiHandler.java:292) ~[classes/:?]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$thenApplyChecked$29(SafeFuture.java:415) ~[classes/:?]
        at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1106) ~[?:?]
        at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2235) ~[?:?]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.thenCompose(SafeFuture.java:463) ~[classes/:?]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.thenApplyChecked(SafeFuture.java:412) ~[classes/:?]
        at tech.pegasys.teku.validator.coordinator.ValidatorApiHandler.lambda$createUnsignedBlock$13(ValidatorApiHandler.java:290) ~[classes/:?]
        at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1072) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073) ~[?:?]
        at tech.pegasys.teku.dataproviders.generators.CachingTaskQueue.lambda$completePendingTask$9(CachingTaskQueue.java:204) ~[classes/:?]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.fromRunnable(SafeFuture.java:154) ~[classes/:?]
        at tech.pegasys.teku.infrastructure.async.AsyncRunner.lambda$runAsync$0(AsyncRunner.java:23) ~[classes/:?]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.of(SafeFuture.java:73) ~[classes/:?]
        at tech.pegasys.teku.infrastructure.async.ScheduledExecutorAsyncRunner.lambda$createRunnableForAction$1(ScheduledExecutorAsyncRunner.java:119) ~[classes/:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:834) [?:?]
Caused by: tech.pegasys.teku.spec.logic.common.statetransition.exceptions.BlockProcessingException: Invalid sync committee signature in SyncAggregate{sync_committee_bits=SszBitvector{size=32, 11111111111111111111111111111111}, sync_committee_signature=SszByteVector{0x8a68f4032257182582d02afe12569a461296d06dc85115d504ac2929e6b277405f3bbac0e730274eadca97747d7ef899030f7f9033efb746dae154c51c1a9a4f088438db8ab15fcf03620d301fc5d21c0ca2af08e8a7d9fafe9864ec8640b318}}
        at tech.pegasys.teku.spec.logic.versions.altair.block.BlockProcessorAltair.processSyncCommittee(BlockProcessorAltair.java:223) ~[classes/:?]
        at tech.pegasys.teku.spec.logic.versions.altair.block.BlockProcessorAltair.processBlock(BlockProcessorAltair.java:99) ~[classes/:?]
        at tech.pegasys.teku.spec.logic.common.block.BlockProcessor.lambda$processBlock$0(BlockProcessor.java:78) ~[classes/:?]
        at tech.pegasys.teku.spec.datastructures.state.beaconstate.common.AbstractBeaconState.updated(AbstractBeaconState.java:59) ~[classes/:?]
        at tech.pegasys.teku.spec.logic.common.block.BlockProcessor.processBlock(BlockProcessor.java:78) ~[classes/:?]
        at tech.pegasys.teku.spec.logic.common.block.BlockProcessor.processBlock(BlockProcessor.java:72) ~[classes/:?]
        at tech.pegasys.teku.spec.logic.common.util.BlockProposalUtil.createNewUnsignedBlock(BlockProposalUtil.java:73) ~[classes/:?]
        ... 21 more
```

on master at `872557e8d1cc77014263604d3839fcb483bdc40e`

--- a/ethereum/core/src/testFixtures/java/tech/pegasys/teku/core/ChainBuilder.java
+++ b/ethereum/core/src/testFixtures/java/tech/pegasys/teku/core/ChainBuilder.java
@@ -37,7 +37,6 @@ import tech.pegasys.teku.core.synccomittee.SignedContributionAndProofTestBuilder
 import tech.pegasys.teku.infrastructure.unsigned.UInt64;
 import tech.pegasys.teku.spec.Spec;
 import tech.pegasys.teku.spec.SpecMilestone;
-import tech.pegasys.teku.spec.SpecVersion;
 import tech.pegasys.teku.spec.TestSpecFactory;
 import tech.pegasys.teku.spec.config.SpecConfigAltair;
 import tech.pegasys.teku.spec.datastructures.blocks.BeaconBlock;
@@ -431,12 +430,15 @@ public class ChainBuilder {
   }
 
   public SignedContributionAndProofTestBuilder createValidSignedContributionAndProofBuilder() {
+    return createValidSignedContributionAndProofBuilder(getLatestSlot());
+  }
+
+  public SignedContributionAndProofTestBuilder createValidSignedContributionAndProofBuilder(
+      final UInt64 slot) {
+    final SyncCommitteeUtil syncCommitteeUtil = spec.getSyncCommitteeUtilRequired(slot);
     final SignedBlockAndState latestBlockAndState = getLatestBlockAndState();
     final Bytes32 beaconBlockRoot = latestBlockAndState.getRoot();
-    final UInt64 slot = latestBlockAndState.getSlot();
-    final UInt64 epoch = spec.computeEpochAtSlot(slot);
-    final SpecVersion specVersion = spec.atSlot(slot);
-    final SyncCommitteeUtil syncCommitteeUtil = specVersion.getSyncCommitteeUtil().orElseThrow();
+    final UInt64 epoch = syncCommitteeUtil.getEpochForDutiesAtSlot(slot);
 
     final Map<UInt64, SyncSubcommitteeAssignments> subcommitteeAssignments =
         syncCommitteeUtil.getSyncSubcommittees(latestBlockAndState.getState(), epoch);

--- a/ethereum/core/src/testFixtures/java/tech/pegasys/teku/core/synccomittee/SignedContributionAndProofTestBuilder.java
+++ b/ethereum/core/src/testFixtures/java/tech/pegasys/teku/core/synccomittee/SignedContributionAndProofTestBuilder.java
@@ -149,7 +149,7 @@ public class SignedContributionAndProofTestBuilder {
       final UInt64 validatorIndex, final Signer signer) {
     final SyncSubcommitteeAssignments assignments =
         syncCommitteeUtil
-            .getSyncSubcommittees(state, spec.computeEpochAtSlot(slot))
+            .getSyncSubcommittees(state, syncCommitteeUtil.getEpochForDutiesAtSlot(slot))
             .get(validatorIndex);
     checkArgument(
         assignments != null && assignments.getAssignedSubcommittees().contains(subcommitteeIndex),

--- a/ethereum/spec/src/main/java/tech/pegasys/teku/spec/datastructures/operations/versions/altair/ValidateableSyncCommitteeSignature.java
+++ b/ethereum/spec/src/main/java/tech/pegasys/teku/spec/datastructures/operations/versions/altair/ValidateableSyncCommitteeSignature.java
@@ -21,6 +21,7 @@ import tech.pegasys.teku.infrastructure.unsigned.UInt64;
 import tech.pegasys.teku.spec.Spec;
 import tech.pegasys.teku.spec.datastructures.state.beaconstate.BeaconState;
 import tech.pegasys.teku.spec.datastructures.util.SyncSubcommitteeAssignments;
+import tech.pegasys.teku.spec.logic.common.util.SyncCommitteeUtil;
 
 public class ValidateableSyncCommitteeSignature {
   private final SyncCommitteeSignature signature;
@@ -61,10 +62,13 @@ public class ValidateableSyncCommitteeSignature {
     if (currentValue.isPresent()) {
       return currentValue.get();
     }
+    final UInt64 signatureSlot = signature.getSlot();
+    final SyncCommitteeUtil syncCommitteeUtil = spec.getSyncCommitteeUtilRequired(signatureSlot);
     final SyncSubcommitteeAssignments assignments =
-        spec.getSyncCommitteeUtilRequired(signature.getSlot())
-            .getSubcommitteeAssignments(
-                state, spec.computeEpochAtSlot(signature.getSlot()), signature.getValidatorIndex());
+        syncCommitteeUtil.getSubcommitteeAssignments(
+            state,
+            syncCommitteeUtil.getEpochForDutiesAtSlot(signatureSlot),
+            signature.getValidatorIndex());
 
     this.subcommitteeAssignments = Optional.of(assignments);
     return assignments;

--- a/ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/util/SyncCommitteeUtil.java
+++ b/ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/util/SyncCommitteeUtil.java
@@ -14,7 +14,6 @@
 package tech.pegasys.teku.spec.logic.common.util;
 
 import static com.google.common.base.Preconditions.checkArgument;
-import static java.util.Collections.emptySet;
 import static java.util.stream.Collectors.toSet;
 import static java.util.stream.Collectors.toUnmodifiableMap;
 import static tech.pegasys.teku.spec.constants.NetworkConstants.SYNC_COMMITTEE_SUBNET_COUNT;
@@ -181,13 +180,6 @@ public class SyncCommitteeUtil {
     return bytesToUInt64(Hash.sha2_256(signature.toSSZBytes()).slice(0, 8)).mod(modulo).isZero();
   }
 
-  public Set<Integer> getSyncSubcommittees(
-      final BeaconState state, final UInt64 epoch, final UInt64 validatorIndex) {
-    final SyncSubcommitteeAssignments assignments =
-        getSyncSubcommittees(state, epoch).get(validatorIndex);
-    return assignments != null ? assignments.getAssignedSubcommittees() : emptySet();
-  }
-
   public Set<Integer> getCommitteeIndices(
       final BeaconState state, final UInt64 epoch, final UInt64 validatorIndex) {
     final SyncSubcommitteeAssignments assignments =
@@ -278,6 +270,21 @@ public class SyncCommitteeUtil {
   public SignedContributionAndProof createSignedContributionAndProof(
       final ContributionAndProof message, final BLSSignature signature) {
     return schemaDefinitionsAltair.getSignedContributionAndProofSchema().create(message, signature);
+  }
+
+  /**
+   * Returns the epoch that should be used for determining sync committee duties at a given slot.
+   *
+   * <p>Sync committee duties are a bit weird, duties should be performed from one slot before the
+   * start of the sync committee period, to one slot before the end. This is so that when the
+   * signatures wind up included in a block at the next slot, it is inside the target sync committee
+   * period.
+   *
+   * @param slot the slot to perform duties for
+   * @return the epoch to use when calculating which duties to perform
+   */
+  public UInt64 getEpochForDutiesAtSlot(final UInt64 slot) {
+    return miscHelpers.computeEpochAtSlot(slot.plus(1));
   }
 
   public UInt64 getMinEpochForSyncCommitteeAssignments(final UInt64 requiredEpoch) {

--- a/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/synccommittee/SignedContributionAndProofValidator.java
+++ b/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/synccommittee/SignedContributionAndProofValidator.java
@@ -149,7 +149,8 @@ public class SignedContributionAndProofValidator {
           contributionAndProof.getAggregatorIndex());
       return REJECT;
     }
-    final UInt64 contributionEpoch = spec.computeEpochAtSlot(contribution.getSlot());
+    final UInt64 contributionEpoch =
+        syncCommitteeUtil.getEpochForDutiesAtSlot(contribution.getSlot());
 
     // [REJECT] The aggregator's validator index is within the current sync subcommittee
     // i.e. state.validators[aggregate_and_proof.aggregator_index].pubkey in

--- a/ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/synccommittee/SignedContributionAndProofValidatorTest.java
+++ b/ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/synccommittee/SignedContributionAndProofValidatorTest.java
@@ -29,6 +29,7 @@ import tech.pegasys.teku.spec.TestSpecFactory;
 import tech.pegasys.teku.spec.config.SpecConfigAltair;
 import tech.pegasys.teku.spec.constants.NetworkConstants;
 import tech.pegasys.teku.spec.datastructures.operations.versions.altair.SignedContributionAndProof;
+import tech.pegasys.teku.spec.logic.common.util.SyncCommitteeUtil;
 import tech.pegasys.teku.spec.util.DataStructureUtil;
 import tech.pegasys.teku.statetransition.validation.InternalValidationResult;
 import tech.pegasys.teku.storage.storageSystem.InMemoryStorageSystemBuilder;
@@ -39,7 +40,7 @@ class SignedContributionAndProofValidatorTest {
   private final SpecConfigAltair config = SpecConfigAltair.required(spec.getGenesisSpecConfig());
   private final DataStructureUtil dataStructureUtil = new DataStructureUtil(spec);
   private final StorageSystem storageSystem =
-      InMemoryStorageSystemBuilder.create().specProvider(spec).numberOfValidators(5).build();
+      InMemoryStorageSystemBuilder.create().specProvider(spec).numberOfValidators(10).build();
   private final ChainBuilder chainBuilder = storageSystem.chainBuilder();
 
   private final SignedContributionAndProofValidator validator =
@@ -57,6 +58,29 @@ class SignedContributionAndProofValidatorTest {
   void shouldAcceptWhenValid() {
     final SignedContributionAndProof message =
         chainBuilder.createValidSignedContributionAndProofBuilder().build();
+    final SafeFuture<InternalValidationResult> result = validator.validate(message);
+    assertThat(result).isCompletedWithValue(ACCEPT);
+  }
+
+  @Test
+  void shouldAcceptWhenValidInSlotLastSlotOfSyncCommitteePeriod() {
+    final SyncCommitteeUtil syncCommitteeUtil = spec.getSyncCommitteeUtilRequired(UInt64.ZERO);
+    final UInt64 period2StartEpoch =
+        syncCommitteeUtil.computeFirstEpochOfNextSyncCommitteePeriod(UInt64.ZERO);
+    final UInt64 period3StartEpoch =
+        syncCommitteeUtil.computeFirstEpochOfNextSyncCommitteePeriod(period2StartEpoch);
+    final UInt64 period2StartSlot = spec.computeStartSlotAtEpoch(period2StartEpoch);
+    final UInt64 period3StartSlot = spec.computeStartSlotAtEpoch(period3StartEpoch);
+    final UInt64 lastSlotOfPeriod = period3StartSlot.minus(1);
+
+    // The first two sync committees are the same so advance the chain into the second period
+    // so we can test going into the third period which is actually different
+    storageSystem.chainUpdater().advanceChainUntil(period2StartSlot);
+    storageSystem.chainUpdater().setCurrentSlot(lastSlotOfPeriod);
+    // Contributions from the last slot of the sync committee period should be valid according to
+    // the next sync committee since that's when they'll be included in blocks
+    final SignedContributionAndProof message =
+        chainBuilder.createValidSignedContributionAndProofBuilder(lastSlotOfPeriod).build();
     final SafeFuture<InternalValidationResult> result = validator.validate(message);
     assertThat(result).isCompletedWithValue(ACCEPT);
   }

--- a/ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/synccommittee/SyncCommitteeSignatureValidatorTest.java
+++ b/ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/synccommittee/SyncCommitteeSignatureValidatorTest.java
@@ -34,6 +34,7 @@ import tech.pegasys.teku.spec.datastructures.operations.versions.altair.Validate
 import tech.pegasys.teku.spec.datastructures.state.beaconstate.versions.altair.BeaconStateAltair;
 import tech.pegasys.teku.spec.datastructures.type.SszPublicKey;
 import tech.pegasys.teku.spec.datastructures.util.SyncSubcommitteeAssignments;
+import tech.pegasys.teku.spec.logic.common.util.SyncCommitteeUtil;
 import tech.pegasys.teku.spec.util.DataStructureUtil;
 import tech.pegasys.teku.storage.client.RecentChainData;
 import tech.pegasys.teku.storage.storageSystem.InMemoryStorageSystemBuilder;
@@ -73,6 +74,39 @@ class SyncCommitteeSignatureValidatorTest {
                 chainBuilder.getLatestBlockAndState().getState(),
                 chainBuilder.getLatestEpoch(),
                 signature.getValidatorIndex());
+    final int validSubnetId = assignments.getAssignedSubcommittees().iterator().next();
+    final ValidateableSyncCommitteeSignature validateableSignature =
+        ValidateableSyncCommitteeSignature.fromNetwork(signature, validSubnetId);
+
+    assertThat(validator.validate(validateableSignature)).isCompletedWithValue(ACCEPT);
+    // Should store the computed subcommittee assignments for the validator.
+    assertThat(validateableSignature.getSubcommitteeAssignments()).contains(assignments);
+  }
+
+  @Test
+  void shouldAcceptWhenValidInSlotLastSlotOfSyncCommitteePeriod() {
+    final SyncCommitteeUtil syncCommitteeUtil = spec.getSyncCommitteeUtilRequired(UInt64.ZERO);
+    final UInt64 period2StartEpoch =
+        syncCommitteeUtil.computeFirstEpochOfNextSyncCommitteePeriod(UInt64.ZERO);
+    final UInt64 period3StartEpoch =
+        syncCommitteeUtil.computeFirstEpochOfNextSyncCommitteePeriod(period2StartEpoch);
+    final UInt64 period2StartSlot = spec.computeStartSlotAtEpoch(period2StartEpoch);
+    final UInt64 period3StartSlot = spec.computeStartSlotAtEpoch(period3StartEpoch);
+    final UInt64 lastSlotOfPeriod = period3StartSlot.minus(1);
+
+    // The first two sync committees are the same so advance the chain into the second period
+    // so we can test going into the third period which is actually different
+    final SignedBlockAndState chainHead =
+        storageSystem.chainUpdater().advanceChainUntil(period2StartSlot);
+    storageSystem.chainUpdater().setCurrentSlot(lastSlotOfPeriod);
+
+    final SyncCommitteeSignature signature =
+        chainBuilder.createSyncCommitteeSignature(lastSlotOfPeriod, chainHead.getRoot());
+    final SyncSubcommitteeAssignments assignments =
+        syncCommitteeUtil.getSubcommitteeAssignments(
+            chainHead.getState(),
+            syncCommitteeUtil.getEpochForDutiesAtSlot(lastSlotOfPeriod),
+            signature.getValidatorIndex());
     final int validSubnetId = assignments.getAssignedSubcommittees().iterator().next();
     final ValidateableSyncCommitteeSignature validateableSignature =
         ValidateableSyncCommitteeSignature.fromNetwork(signature, validSubnetId);

--- a/networking/eth2/src/test/java/tech/pegasys/teku/networking/eth2/gossip/SyncCommitteeSignatureGossipManagerTest.java
+++ b/networking/eth2/src/test/java/tech/pegasys/teku/networking/eth2/gossip/SyncCommitteeSignatureGossipManagerTest.java
@@ -97,10 +97,12 @@ class SyncCommitteeSignatureGossipManagerTest {
         ValidateableSyncCommitteeSignature.fromValidator(
             dataStructureUtil.randomSyncCommitteeSignature());
 
+    final UInt64 dutyEpoch = UInt64.valueOf(333);
     final BeaconStateAltair state = dataStructureUtil.stateBuilderAltair().build();
     when(syncCommitteeStateUtils.getStateForSyncCommittee(
             signature.getSlot(), signature.getBeaconBlockRoot()))
         .thenReturn(SafeFuture.completedFuture(Optional.of(state)));
+    when(syncCommitteeUtil.getEpochForDutiesAtSlot(any())).thenReturn(dutyEpoch);
     when(syncCommitteeUtil.getSubcommitteeAssignments(any(), any(), any()))
         .thenReturn(
             SyncSubcommitteeAssignments.builder()
@@ -112,8 +114,7 @@ class SyncCommitteeSignatureGossipManagerTest {
     gossipManager.publish(signature);
 
     verify(syncCommitteeUtil)
-        .getSubcommitteeAssignments(
-            state, UInt64.ZERO, signature.getSignature().getValidatorIndex());
+        .getSubcommitteeAssignments(state, dutyEpoch, signature.getSignature().getValidatorIndex());
     verify(subnetSubscriptions).gossip(signature.getSignature(), 1);
     verify(subnetSubscriptions).gossip(signature.getSignature(), 3);
     verify(subnetSubscriptions).gossip(signature.getSignature(), 5);

--- a/validator/client/src/main/java/tech/pegasys/teku/validator/client/SyncCommitteeScheduler.java
+++ b/validator/client/src/main/java/tech/pegasys/teku/validator/client/SyncCommitteeScheduler.java
@@ -63,13 +63,13 @@ public class SyncCommitteeScheduler implements ValidatorTimingChannel {
       return;
     }
     final SyncCommitteeUtil syncCommitteeUtil = maybeUtils.get();
-    final UInt64 currentEpoch = spec.computeEpochAtSlot(slot);
-    final SpecConfigAltair specConfig = SpecConfigAltair.required(spec.getSpecConfig(currentEpoch));
+    final UInt64 dutiesEpoch = syncCommitteeUtil.getEpochForDutiesAtSlot(slot);
+    final SpecConfigAltair specConfig = SpecConfigAltair.required(spec.getSpecConfig(dutiesEpoch));
     if (currentSyncCommitteePeriod.isEmpty()) {
       final SyncCommitteePeriod committeePeriod =
           createSyncCommitteePeriod(
               syncCommitteeUtil,
-              syncCommitteeUtil.computeFirstEpochOfCurrentSyncCommitteePeriod(currentEpoch),
+              syncCommitteeUtil.computeFirstEpochOfCurrentSyncCommitteePeriod(dutiesEpoch),
               0);
       committeePeriod.calculateDuties();
       currentSyncCommitteePeriod = Optional.of(committeePeriod);
@@ -77,7 +77,7 @@ public class SyncCommitteeScheduler implements ValidatorTimingChannel {
 
     if (nextSyncCommitteePeriod.isEmpty()) {
       final UInt64 firstEpochOfNextSyncCommitteePeriod =
-          syncCommitteeUtil.computeFirstEpochOfNextSyncCommitteePeriod(currentEpoch);
+          syncCommitteeUtil.computeFirstEpochOfNextSyncCommitteePeriod(dutiesEpoch);
       final int subscribeEpochsPriorToNextSyncPeriod =
           earlySubscribeRandomSource.nextInt(specConfig.getEpochsPerSyncCommitteePeriod());
       nextSyncCommitteePeriod =
@@ -89,10 +89,10 @@ public class SyncCommitteeScheduler implements ValidatorTimingChannel {
     }
 
     final SyncCommitteePeriod nextSyncCommitteePeriod = this.nextSyncCommitteePeriod.get();
-    if (currentEpoch.isGreaterThanOrEqualTo(nextSyncCommitteePeriod.subscribeEpoch)) {
+    if (dutiesEpoch.isGreaterThanOrEqualTo(nextSyncCommitteePeriod.subscribeEpoch)) {
       nextSyncCommitteePeriod.calculateDuties();
     }
-    if (currentEpoch.isGreaterThanOrEqualTo(nextSyncCommitteePeriod.periodStartEpoch)) {
+    if (dutiesEpoch.isGreaterThanOrEqualTo(nextSyncCommitteePeriod.periodStartEpoch)) {
       this.currentSyncCommitteePeriod = this.nextSyncCommitteePeriod;
       this.nextSyncCommitteePeriod = Optional.empty();
     }
@@ -119,7 +119,12 @@ public class SyncCommitteeScheduler implements ValidatorTimingChannel {
   }
 
   private Optional<PendingDuties> getDutiesForSlot(final UInt64 slot) {
-    final UInt64 epoch = spec.computeEpochAtSlot(slot);
+    final Optional<SyncCommitteeUtil> maybeUtils = spec.getSyncCommitteeUtil(slot);
+    if (maybeUtils.isEmpty()) {
+      return Optional.empty();
+    }
+    final SyncCommitteeUtil syncCommitteeUtil = maybeUtils.get();
+    final UInt64 epoch = syncCommitteeUtil.getEpochForDutiesAtSlot(slot);
     return Stream.of(currentSyncCommitteePeriod, nextSyncCommitteePeriod)
         .flatMap(Optional::stream)
         .filter(period -> period.isCurrentPeriodForEpoch(epoch))

--- a/validator/client/src/test/java/tech/pegasys/teku/validator/client/SyncCommitteeSchedulerTest.java
+++ b/validator/client/src/test/java/tech/pegasys/teku/validator/client/SyncCommitteeSchedulerTest.java
@@ -160,7 +160,7 @@ class SyncCommitteeSchedulerTest {
   }
 
   @Test
-  void shouldSwitchToNextCommitteePeriodWhenFirstSlotReached() {
+  void shouldSwitchToNextCommitteePeriodWhenLastSlotOfSyncCommitteePeriodReached() {
     when(earlySubscribeRandomSource.nextInt(epochsPerSyncCommitteePeriod)).thenReturn(5);
     final UInt64 nextSyncCommitteePeriodStartEpoch =
         syncCommitteeUtil.computeFirstEpochOfNextSyncCommitteePeriod(UInt64.ZERO);
@@ -182,10 +182,11 @@ class SyncCommitteeSchedulerTest {
     scheduler.onAttestationCreationDue(subscribeSlot);
     verify(duties).performProductionDuty(subscribeSlot);
 
-    scheduler.onSlot(nextSyncCommitteePeriodStartSlot);
-    scheduler.onAttestationCreationDue(nextSyncCommitteePeriodStartSlot);
-    verify(nextDuties).performProductionDuty(nextSyncCommitteePeriodStartSlot);
-    verify(duties, never()).performProductionDuty(nextSyncCommitteePeriodStartSlot);
+    final UInt64 nextSyncCommitteeFirstDutySlot = nextSyncCommitteePeriodStartSlot.minus(1);
+    scheduler.onSlot(nextSyncCommitteeFirstDutySlot);
+    scheduler.onAttestationCreationDue(nextSyncCommitteeFirstDutySlot);
+    verify(nextDuties).performProductionDuty(nextSyncCommitteeFirstDutySlot);
+    verify(duties, never()).performProductionDuty(nextSyncCommitteeFirstDutySlot);
   }
 
   /**

